File: library/src/main/java/org/xmtp/android/library/libxmtp/DecodedMessage.kt
package org.xmtp.android.library.libxmtp

import android.util.Log
import org.xmtp.android.library.InboxId
import org.xmtp.android.library.XMTPException
import org.xmtp.android.library.codecs.ContentTypeGroupUpdated
import org.xmtp.android.library.codecs.EncodedContent
import org.xmtp.android.library.codecs.decoded
import org.xmtp.android.library.codecs.id
import org.xmtp.android.library.messages.Topic
import org.xmtp.android.library.toHex
import org.xmtp.proto.message.contents.Content
import uniffi.xmtpv3.FfiConversationMessageKind
import uniffi.xmtpv3.FfiDeliveryStatus
import uniffi.xmtpv3.FfiMessage
import uniffi.xmtpv3.FfiMessageWithReactions
import java.util.Date

class DecodedMessage private constructor(
    private val libXMTPMessage: FfiMessage,
    val encodedContent: Content.EncodedContent,
    private val decodedContent: Any?,
    val childMessages: List<DecodedMessage>? = null
) {
    enum class MessageDeliveryStatus {
        ALL, PUBLISHED, UNPUBLISHED, FAILED
    }

    enum class SortDirection {
        ASCENDING,
        DESCENDING;
    }

    val id: String
        get() = libXMTPMessage.id.toHex()

    val conversationId: String
        get() = libXMTPMessage.conversationId.toHex()

    val senderInboxId: InboxId
        get() = libXMTPMessage.senderInboxId

    val sentAt: Date
        get() = Date(libXMTPMessage.sentAtNs / 1_000_000)

    val sentAtNs: Long
        get() = libXMTPMessage.sentAtNs

    val deliveryStatus: MessageDeliveryStatus
        get() = when (libXMTPMessage.deliveryStatus) {
            FfiDeliveryStatus.UNPUBLISHED -> MessageDeliveryStatus.UNPUBLISHED
            FfiDeliveryStatus.PUBLISHED -> MessageDeliveryStatus.PUBLISHED
            FfiDeliveryStatus.FAILED -> MessageDeliveryStatus.FAILED
        }

    val topic: String
        get() = Topic.groupMessage(conversationId).description

    @Suppress("UNCHECKED_CAST")
    fun <T> content(): T? = decodedContent as? T

    val fallback: String
        get() = encodedContent.fallback

    val body: String
        get() {
            return content() as? String ?: fallback
        }

    companion object {
        fun create(libXMTPMessage: FfiMessage): DecodedMessage? {
            return try {
                val encodedContent = EncodedContent.parseFrom(libXMTPMessage.content)
                Log.d("XMTP Message Create", "encodedContent type:" + encodedContent.type.id)
                if (encodedContent.type == ContentTypeGroupUpdated && libXMTPMessage.kind != FfiConversationMessageKind.MEMBERSHIP_CHANGE) {
                    throw XMTPException("Error decoding group membership change")
                }
                // Decode the content once during creation
                val decodedContent = encodedContent.decoded<Any>()
                DecodedMessage(libXMTPMessage, encodedContent, decodedContent)
            } catch (e: Exception) {
                null // Return null if decoding fails
            }
        }

        fun create(libXMTPMessageWithReactions: FfiMessageWithReactions): DecodedMessage? {
            return try {
                val encodedContent = EncodedContent.parseFrom(libXMTPMessageWithReactions.message.content)
                if (encodedContent.type == ContentTypeGroupUpdated && libXMTPMessageWithReactions.message.kind != FfiConversationMessageKind.MEMBERSHIP_CHANGE) {
                    throw XMTPException("Error decoding group membership change")
                }
                // Decode the content once during creation
                val decodedContent = encodedContent.decoded<Any>()

                // Convert reactions to Message objects
                val reactionMessages = libXMTPMessageWithReactions.reactions.mapNotNull { create(it) }

                DecodedMessage(
                    libXMTPMessageWithReactions.message,
                    encodedContent,
                    decodedContent,
                    reactionMessages
                )
            } catch (e: Exception) {
                null // Return null if decoding fails
            }
        }
    }
}

File: Sources/XMTPiOS/Libxmtp/DecodedMessage.swift
import Foundation
import LibXMTP

enum DecodedMessageError: Error {
	case decodeError(String)
}

public enum MessageDeliveryStatus: String, Sendable {
	case all
	case published
	case unpublished
	case failed
}

public enum SortDirection {
	case ascending
	case descending
}

public struct DecodedMessage: Identifiable {
	let ffiMessage: FfiMessage
	private let decodedContent: Any?
	public let childMessages: [DecodedMessage]?

	public var id: String {
		ffiMessage.id.toHex
	}

	public var conversationId: String {
		ffiMessage.conversationId.toHex
	}

	public var senderInboxId: InboxId {
		ffiMessage.senderInboxId
	}

	public var sentAt: Date {
		Date(
			timeIntervalSince1970: TimeInterval(ffiMessage.sentAtNs)
				/ 1_000_000_000)
	}

	public var sentAtNs: Int64 {
		ffiMessage.sentAtNs
	}

	public var deliveryStatus: MessageDeliveryStatus {
		switch ffiMessage.deliveryStatus {
		case .unpublished:
			return .unpublished
		case .published:
			return .published
		case .failed:
			return .failed
		}
	}

	public var topic: String {
		Topic.groupMessage(conversationId).description
	}

	public func content<T>() throws -> T {
		guard let result = decodedContent as? T else {
			throw DecodedMessageError.decodeError(
				"Decoded content could not be cast to the expected type \(T.self)."
			)
		}
		return result
	}

	public var fallback: String {
		get throws {
			try encodedContent.fallback
		}
	}

	public var body: String {
		get throws {
			do {
				return try content() as String
			} catch {
				return try fallback
			}
		}
	}

	public var encodedContent: EncodedContent {
		get throws {
			try EncodedContent(serializedBytes: ffiMessage.content)
		}
	}

	public static func create(ffiMessage: FfiMessage)
		-> DecodedMessage?
	{
		do {
			let encodedContent = try EncodedContent(
				serializedBytes: ffiMessage.content)
			if encodedContent.type == ContentTypeGroupUpdated
				&& ffiMessage.kind != .membershipChange
			{
				throw DecodedMessageError.decodeError(
					"Error decoding group membership change")
			}
			// Decode the content once during creation
			let decodedContent: Any = try encodedContent.decoded()
			return DecodedMessage(
				ffiMessage: ffiMessage, decodedContent: decodedContent,
				childMessages: nil)
		} catch {
			print("Error creating Message: \(error)")
			return nil
		}
	}

	public static func create(ffiMessage: FfiMessageWithReactions)
		-> DecodedMessage?
	{
		do {
			let encodedContent = try EncodedContent(
				serializedBytes: ffiMessage.message.content)
			if encodedContent.type == ContentTypeGroupUpdated
				&& ffiMessage.message.kind != .membershipChange
			{
				throw DecodedMessageError.decodeError(
					"Error decoding group membership change")
			}
			// Decode the content once during creation
			let decodedContent: Any = try encodedContent.decoded()

			let childMessages = try ffiMessage.reactions.map { reaction in
				let encodedContent = try EncodedContent(
					serializedBytes: reaction.content)
				// Decode the content once during creation
				let decodedContent: Any = try encodedContent.decoded()
				return DecodedMessage(
					ffiMessage: reaction, decodedContent: decodedContent,
					childMessages: nil)
			}

			return DecodedMessage(
				ffiMessage: ffiMessage.message, decodedContent: decodedContent,
				childMessages: childMessages)
		} catch {
			print("Error creating Message: \(error)")
			return nil
		}
	}
}
